src/libstd/sys/unix/ext/net.rs

```diff
@@ -0,0 +1,1225 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+#![unstable(feature = "unix_socket", reason = "newly added", issue = "32312")]
+
+//! Unix-specific networking functionality
+
+use libc;
+
+use prelude::v1::*;
+use ascii;
+use ffi::OsStr;
+use fmt;
+use io;
+use mem;
+use net::Shutdown;
+use os::unix::ffi::OsStrExt;
+use os::unix::io::{RawFd, AsRawFd, FromRawFd, IntoRawFd};
+use path::Path;
+use time::Duration;
+use sys::cvt;
+use sys_common::io::read_to_end_uninitialized;
+
+fn sun_path_offset() -> usize {
+    unsafe {
+        // Work with an actual instance of the type since using a null pointer is UB
+        let addr: libc::sockaddr_un = mem::uninitialized();
+        let base = &addr as *const _ as usize;
+        let path = &addr.sun_path as *const _ as usize;
+        path - base
+    }
```

Comment:
The IR we generate for that would be:

```rust

struct Foo {
    a: u32,
    b: u16,
    c: u8,
}

pub fn foo() -> usize {
    unsafe {
        &((*(0 as *mut Foo)).c) as *const _ as usize
    }
}

```

```llvm
define i64 @_ZN3foo20hd1daefe8e1d28080naaE() unnamed_addr #0 {
entry-block:
  ret i64 ptrtoint (i8* getelementptr inbounds (%Foo, %Foo* null, i32 0, i32 2) to i64)
}
```

where the [definition of `inbounds`](http://llvm.org/docs/LangRef.html#id213) is:

> If the inbounds keyword is present, the result value of the getelementptr is a poison value if the base pointer is not an in bounds address of an allocated object, or if any of the addresses that would be formed by successive addition of the offsets implied by the indices to the base address with infinitely precise signed arithmetic are not an in bounds address of that allocated object. The in bounds addresses for an allocated object are all the addresses that point into the object, plus the address one byte past the end. In cases where the base is a vector of pointers the inbounds keyword applies to each of the computations element-wise.

Doesn't seem to mention anything about a null pointer?